### ISSUE TITLE
Regression: type information is not preserved when converting a constructor to a callable

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#8116]
Starting with version 1.1.359 the type hints for dataclass parameters are lost after converting it to a callable.

```python
from collections.abc import Callable
from dataclasses import dataclass

@dataclass
class Dataclass:
    value: int

def f[From, To](c: Callable[[From], To]) -> Callable[[From], To]:
    return c

reveal_type(f(Dataclass))
```

1.1.358:

```
Type of "f(Dataclass)" is "(int) -> Dataclass"
```

1.1.359:

```
Type of "f(Dataclass)" is "(Any) -> Dataclass"
```

**Comment by LeeeeT** [comment]
This also happens with `NewType`.

```python
from collections.abc import Callable
from typing import NewType

C = NewType("C", int)

def f[From, To](c: Callable[[From], To]) -> Callable[[From], To]:
    return c

reveal_type(f(C))
```

```
Type of "f(C)" is "(Any) -> C"
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.367.
